### PR TITLE
Interop 2023 dashboard changes

### PR DESCRIPTION
**Pull request by DanielRyanSmith**
These changes are for the Interop 2023 dashboard redesign.

[Staging environment link here](https://interop-2023-changes-dot-wptdashboard-staging.uk.r.appspot.com/interop-2023). **NOTE**: This staging environment is reading from files generated from the [results-analysis repo](https://github.com/web-platform-tests/results-analysis). If [the changes to the results-analysis script](https://github.com/web-platform-tests/results-analysis/pull/140) have not been merged yet, there is a chance that this view will not be functional.

Explanation of changes by file:
- webapp/components/wpt-header.js: Update navbar to display a link with Interop 2023 rather than 2022.
- webapp/interop_handler.go: Change handler to allow navigation to /interop-2023.
- webapp/static/common.css: Increase page size maximum from 1200px to 1400px.
- webapp/static/interop-data.json: Add all information needed for display of Interop 2023, as well as a new `countsTowardScore` property that is added to focus areas to distinguish if they should be counted toward the overall summary browser and interop scores.

Most major changes come in webapp/components/interop.js. An explanation of those changes will accompany screenshots.

---

The interop 2023 page will be displayed in accordance to the new redesign.
Screenshot 2023-01-22 at 5 53 05 PM

Interop 2023 will display an additional explainer underneath the graph. This is exclusive to the 2023 dashboard.
Screenshot 2023-01-22 at 5 53 42 PM

The Interop 2022 dashboard will display with some new aspects of the redesign, including the interop and investigation summary scores. However, it will still show a block-level display.
Screenshot 2023-01-22 at 5 54 12 PM

Interop 2021 (or Compat 2021) will also display with some stylistic aspects of the new redesign, but will not display investigation or summary number bubbles. The interop score will be available in the table, however.
Screenshot 2023-01-22 at 5 54 36 PM

The site is not optimized for mobile viewing, however, a functional view will be available when viewing on mobile. 
Screenshot 2023-01-22 at 5 55 08 PM

### CODE REVIEW

**Comment by DanielRyanSmith** [comment]
@nt1m I tried adding the viewport meta tag as suggested, but the mobile layout devolved into something that looked more incorrect. I have to assume it is due to some of my incorrect style decisions or an approach, but I am thinking it might be best to ship this with the site's existing mobile functionality and to address this sometime in the future.

Screenshot 2023-01-22 at 6 13 25 PM

**Comment by nt1m** [comment]
> @nt1m I tried adding the viewport meta tag as suggested, but the mobile layout devolved into something that looked more incorrect. I have to assume it is due to some of my incorrect style decisions or an approach, but I am thinking it might be best to ship this with the site's existing mobile functionality and to address this sometime in the future.
> 
> Screenshot 2023-01-22 at 6 13 25 PM

Ah yeah, the bubbles likely need to be smaller. The viewport meta tag was a suggestion in response to the font-size increase you made for mobile, which looked wrong on desktop.

File: webapp/components/interop.js [hunk 1]
```diff
@@ -989,22 +1140,56 @@ class InteropSummary extends PolymerElement {
...
+
+  shouldDisplayInvestigationNumber() {
+    const scores = this.dataManager.getYearProp('investigationScores');
+    return scores !== null && scores !== undefined;
+  }
+
+  // Takes a summary number div and changes the value to match the score (with CountUp).
+  updateSummaryScore(number, score) {
+    score = Math.floor(score / 10);
+    let curScore = number.innerText;
```
Comment (KyleJu) [inline webapp/components/interop.js:1164]:
```suggestion
    const curScore = number.innerText;
```

File: webapp/components/interop.js [hunk 2]
```diff
@@ -989,22 +1140,56 @@ class InteropSummary extends PolymerElement {
...
+  }
+
   async updateSummaryScores() {
-    let numbers = this.shadowRoot.querySelectorAll('.summary-number');
+    const scoreNumbers = this.shadowRoot.querySelectorAll('.score-number');
     const scores = this.stable ? this.scores.stable : this.scores.experimental;
-    if (numbers.length !== scores.length) {
-      throw new Error(`Mismatched number of browsers/scores: ${numbers.length} vs. ${this.scores.length}`);
+    const summaryFeatureName = this.dataManager.getYearProp('summaryFeatureName');
+    if (!scores.length || scoreNumbers.length !== scores.length) {
```
Comment (KyleJu) [inline webapp/components/interop.js:1177]:
```suggestion
    if (!scores.length && scoreNumbers.length !== scores.length) {
```
Do you mean &&?

**Review by KyleJu — COMMENTED** [review]
Pending HTML review

File: webapp/components/interop.js [hunk 3]
```diff
@@ -429,182 +502,228 @@ class InteropDashboard extends PolymerElement {
...
-                <th><b>GRAND TOTAL</b></th>
-                <th>[[totalChromium]]</th>
-                <th>[[totalFirefox]]</th>
-                <th>[[totalSafari]]</th>
-              </tr>
-            </tfoot>
-          </table>
+              </tbody>
+            </table>
+          </div>
```
Comment (DanielRyanSmith) [inline webapp/components/interop.js:688]: This section changes the "SUBTOTAL" text to "TOTAL", and adds a template to display a section for previous investigation scores if any investigation scores are defined for the previous year.

File: webapp/components/interop.js [hunk 4]
```diff
@@ -429,182 +502,228 @@ class InteropDashboard extends PolymerElement {
...
         </div>
-
-        <interop-feature-chart year="[[year]]"
-                               data-manager="[[dataManager]]"
-                               stable="[[stable]]"
-                               feature="{{feature}}">
-        </interop-feature-chart>
-      </section>
+      </div>
       <footer class="compat-footer">
```
Comment (DanielRyanSmith) [inline webapp/components/interop.js:726]: Most of the footer text here has been moved into the the grid's description, and the footer now only displays links to specific years.

File: webapp/components/interop.js [hunk 5]
```diff
@@ -895,75 +1033,88 @@ class InteropSummary extends PolymerElement {
...
+          <!-- Interop -->
+          <div id="interopSummary" class="summary-flex-item" tabindex="0">
+            <h3 class="summary-title">INTEROP</h3>
+            <div class="summary-number score-number">--</div>
+          </div>
+          <!-- Investigations -->
+          <div id="investigationSummary" class="summary-flex-item" tabindex="0">
+            <h3 class="summary-title">INVESTIGATIONS</h3>
+            <div id="investigationNumber" class="summary-number">--</div>
+          </div>
```
Comment (DanielRyanSmith) [inline webapp/components/interop.js:1047]: This section adds the interop and investigation summary bubbles at the top of the page for 2023. The are populated with data similarly to the other summary bubbles. Again, indentations make this look like a larger change.

File: webapp/components/interop.js [hunk 6]
```diff
@@ -34,15 +34,29 @@ class InteropDataManager {
...
+
+    // Calc and save investigation scores.
+    this.investigationScores = yearInfo.investigation_scores;
+    this.investigationWeight = yearInfo.investigation_weight;
+    // If the previous year has an investigation score, save it for later reference.
+    if (paramsByYear[parseInt(this.year) - 1]) {
+      this.previousInvestigationScores = paramsByYear[previousYear].investigation_scores;
+    }
+    if (this.previousInvestigationScores) {
+      this.#calcInvestigationTotalScore(this.previousInvestigationScores, true);
```
[the remark this answers is not in the document]
Comment (DanielRyanSmith) [reply webapp/components/interop.js]: Much more succinct implementation. Good call. 🙂

File: webapp/components/interop.js [hunk 7]
```diff
@@ -120,11 +154,12 @@ class InteropDataManager {
       'Chrome',
       'Firefox',
       'Safari',
+      'Interop',
     ];
-    const focusAreaLabels = Object.keys(this.focusAreas);
+    const focusAreaLabels = this.focusAreasList;
```
[the remark this answers is not in the document]
Comment (DanielRyanSmith) [reply webapp/components/interop.js]: Removed! Not needed.

File: webapp/components/interop.js [hunk 8]
```diff
@@ -165,23 +200,26 @@ class InteropDataManager {
...
+          // counts toward the total browser score.
+          if (this.focusAreas[feature].countsTowardScore) {
+            testScore += score;
+          }
         });
 
-        testScore /= numFocusAreas;
+        // Count up the number of focus areas that count toward the browser score
+        // to handle averaging.
+        const numCountedFocusAreas = this.focusAreasList.reduce(
```
[the remark this answers is not in the document]
Comment (DanielRyanSmith) [reply webapp/components/interop.js]: This looks a bit easier to read than the reduce approach, so I think this is a good change 🙂

File: webapp/components/interop.js [hunk 9]
```diff
@@ -165,23 +200,26 @@ class InteropDataManager {
...
+          (sum, k) => (this.focusAreas[k].countsTowardScore) ? sum + 1 : sum, 0);
+        testScore /= numCountedFocusAreas;
```
[the remark this answers is not in the document]
Comment (DanielRyanSmith) [reply webapp/components/interop.js:221]: Changed to 0.0 👍

File: webapp/components/interop.js [hunk 10]
```diff
@@ -429,182 +502,228 @@ class InteropDashboard extends PolymerElement {
...
+        }
+
+        /* TODO(danielrsmith): This is a workaround to avoid the text scaling that
+         * happens for p tags on mobile, but not for any other text (like the focus area table).
+         * Remove this when deeper mobile functionality has been added. */
+        p {
+          text-size-adjust: none;
+        }
+
+      </style>
```
[the remark this answers is not in the document]
Comment (DanielRyanSmith) [reply webapp/components/interop.js:549]: I added some basic comments on the PR below, but I can give you some further information if needed

File: webapp/components/interop.js (hunk 1 above)
Comment (DanielRyanSmith) [reply webapp/components/interop.js]: Fixed 👍

File: webapp/components/interop.js [hunk 11]
```diff
@@ -684,6 +802,16 @@ class InteropDashboard extends PolymerElement {
 
     this.$.toggleStable.setAttribute('aria-pressed', this.stable);
     this.$.toggleExperimental.setAttribute('aria-pressed', !this.stable);
+    // Keep the block-level design for interop 2021-2022
+    if (this.year !== '2023') {
+      const gridContainerDiv = this.shadowRoot.querySelector('.grid-container');
+      gridContainerDiv.style.display = 'block';
```
[the remark this answers is not in the document]
Comment (DanielRyanSmith) [reply webapp/components/interop.js:807]: > Do these style changes account for the different screen size? I see .grid-container is declared differently based on media type

This is a good questions, and after checking this should still function as expected despite screen resizing. This just ensures that 2021 and 2022 are shown similarly to their existing designs.

> Also optional: maybe we can use Attribute selectors here and declare these style changes under <style>. e.g.
div[2023] {
display: block;
}

I'm not familiar with attribute selection, but the biggest concern is setting the style conditionally based on which year we're viewing, and I haven't studied how to handle that specifically.

File: webapp/components/interop.js (hunk 2 above)
Comment (DanielRyanSmith) [reply webapp/components/interop.js]: This is actually from an older implementation of adding the investigation score and the first part of the logic should not be in the statement anymore. Thanks for the catch 🙂

File: webapp/interop_handler.go [hunk 12]
```diff
@@ -18,10 +18,11 @@ type interopData struct {
 }
 
 // Set of years that are valid for Interop 20XX.
-var validYears = map[string]bool{"2021": true, "2022": true}
+var validYears = map[string]bool{"2021": true, "2022": true, "2023": true}
 
 // Year that any invalid year will redirect to.
-const defaultRedirectYear = "2022"
+// TODO(danielrsmith): Change this redirect for next year's interop page.
```
Comment (jcscottiii) [inline webapp/interop_handler.go:24]: Should create a bug for Interop 2024 that mentions the need to change this or some other artifact to remind future self that you need to change this. Not a blocker for this PR

Comment (DanielRyanSmith) [reply webapp/interop_handler.go:24]: That's a good idea, as it's only capured in the code here. I'll make one!

**Review by jcscottiii — APPROVED** [review]
LGTM. Added some suggestions.

Other suggestions but not needed for this PR:
- split up interop.js into smaller components.
- A test to check the structure of interop-data.json. You could create a class that it is strongly coupled to. That way when you read it in, you know the structure.

**Comment by nt1m** [comment]
Some comments on the dashboard:
* Color spaces category for 2023 should link to a query with: `(label:interop-2022-color or label:interop-2023-color)`
* Ditto for Forms
* Grid/Flexbox for 2023 should link to a query with `(label:interop-2021-flexbox or label:interop-2023-flexbox)`
* Graph colors are not colorblind friendly (you can use the Firefox accessibility tool to test), the lighter green recommended in the document should address that (although not sure it would be differentiable from the Interop yellow)

File: webapp/static/interop-data.json [hunk 13]
```diff
@@ -413,7 +413,7 @@
                 "description": "Forms",
                 "mdn": "https://developer.mozilla.org/docs/Web/HTML/Element/form",
                 "spec": "https://html.spec.whatwg.org/multipage/forms.html#the-form-element",
-                "tests": "/results/?label=experimental&label=master&product=chrome&product=firefox&product=safari&aligned&view=interop&q=label%3Ainterop-2023-forms",
+                "tests": "/results/?label=master&label=experimental&product=chrome&product=firefox&product=safari&aligned&view=interop&q=label%3Ainterop-2022-forms%20or%20label%3Ainterop-2023-forms",
                 "countsTowardScore": true
             },
             "interop-2023-grid": {
```
Comment (nt1m) [inline webapp/static/interop-data.json:419]: Grid needs the same treatment as Flexbox too

File: webapp/components/interop.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (KyleJu) [reply webapp/components/interop.js:807]: Thanks for checking. We can come back to the second suggestion later

**Review by KyleJu — APPROVED** [review]
LGTM. Also to second what James said, we should split this file to smaller/reusable Polymer components as it is getting harder to review. We can do that in a follow-up PR

File: webapp/components/interop.js [hunk 14]
```diff
@@ -56,6 +72,22 @@ class InteropDataManager {
       this.experimentalDatatables.get(feature);
   }
 
+  // Calculates the over investigation score to be displayed in the summary bubble
```
Comment (KyleJu) [inline webapp/components/interop.js:75]:
```suggestion
  // Calculates the investigation score to be displayed in the summary bubble
```

File: webapp/components/interop.js [hunk 15]
```diff
@@ -56,6 +72,22 @@ class InteropDataManager {
...
+      return undefined;
+    }
+    // Get the last listed score for each category and sum them.
+    const totalScore = investigationScores.reduce((sum, area) => {
+      if (area.scores_over_time.length > 0) {
+        return sum + area.scores_over_time[area.scores_over_time.length - 1].score;
+      }
+      return sum;
+    }, 0);
+    return totalScore / investigationScores.length;
```
Comment (KyleJu) [inline webapp/components/interop.js:88]: Nit: if float is needed here, we should do totalScore * 1.0 / ...

File: webapp/components/interop.js [hunk 16]
```diff
@@ -989,22 +1139,56 @@ class InteropSummary extends PolymerElement {
...
   async updateSummaryScores() {
-    let numbers = this.shadowRoot.querySelectorAll('.summary-number');
+    const scoreNumbers = this.shadowRoot.querySelectorAll('.score-number');
     const scores = this.stable ? this.scores.stable : this.scores.experimental;
-    if (numbers.length !== scores.length) {
-      throw new Error(`Mismatched number of browsers/scores: ${numbers.length} vs. ${this.scores.length}`);
+    const summaryFeatureName = this.dataManager.getYearProp('summaryFeatureName');
+    if (scoreNumbers.length !== scores.length) {
+      throw new Error(`Mismatched number of browsers/scores: 
+${scoreNumbers.length} vs. ${scores.length}`);
```
Comment (KyleJu) [inline webapp/components/interop.js:1178]: The indentation needs to be fixed

File: webapp/components/interop.js [hunk 17]
```diff
@@ -1251,15 +1396,16 @@ class InteropFeatureChart extends PolymerElement {
...
+      colors: ['#7be83f', '#F57400', '#0095F0', '#FCBA2F'],
     };
 
     // We draw the chart in two ways, depending on the viewport width. In
     // 'full' mode the legend is on the right and we limit the chart size to
     // 700px wide. In 'mobile' mode the legend is on the top and we use all the
     // space we can get for the chart.
     if (containerDiv.clientWidth >= 700) {
-      options.width = 700;
+      options.width = '100%';
```
Comment (KyleJu) [inline webapp/components/interop.js:1408]: Is the comment described above still correct (we limit the chart size to
    // 700px wide)?  If not, maybe we should update the comment.

File: webapp/static/interop-data.json (hunk 13 above)
Comment (DanielRyanSmith) [reply webapp/static/interop-data.json:419]: Done!

File: webapp/components/interop.js (hunk 17 above)
Comment (DanielRyanSmith) [reply webapp/components/interop.js]: updated for clarity

**Comment by DanielRyanSmith** [comment]
Thanks everyone! Definitely agree that this file needs componentized. I'll file an issue.

**Pull request merged**